jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -36,23 +43,61 @@
   /** Helper for loading a .class file. */
   private static class ClassFileLoader extends ClassLoader {
 
-    private final Path classFile;
+    private static final Map<Path, Class<?>> cache = new ConcurrentHashMap<>();
 
-    private ClassFileLoader(Path classFile) {
-      this.classFile = classFile;
+    private final Path rootDirectory;
+    private final Path defaultClassFile;
+
+    private ClassFileLoader(Path defaultClassFile, Path rootDirectory) {
+      this.defaultClassFile = defaultClassFile;
+      this.rootDirectory = rootDirectory;
     }
 
     @Nullable
     @Override
     public Class findClass(@Nullable String name) {
       try {
-        byte[] bytes = Files.readAllBytes(classFile);
-        return defineClass(name, bytes, 0, bytes.length);
-      } catch (IOException | ClassFormatError ignored) {
+        // Name is only ever null when we call findClass manually, and not null otherwise. If null,
+        // we should resolve the correct filename.
+        Path file = (name == null ? defaultClassFile : getPathFromClassName(name));
+        if (cache.containsKey(file)) {
+          return cache.get(file);
+        }
+
+        if (!Files.exists(file)) {
+          // TODO: Log search class failure?
+          return null;
+        }
+
+        byte[] bytes = Files.readAllBytes(file);
+        Class<?> definedClass = defineClass(name, bytes, 0, bytes.length);
+        cache.put(file, definedClass);
+        return definedClass;
+
+      } catch (IOException
+          | ClassFormatError
+          | IndexOutOfBoundsException
+          | SecurityException
+          | NoClassDefFoundError ignored) {
         // Not a valid class file
+        // TODO: Log search class failure when NoClassDefFoundError/SecurityException is caught?
         return null;
       }
     }
+
+    /**
+     * Converts a class name (pack.age.ClassName) to a Path (rootFolder/pack/age/ClassName.class).
+     */
+    private Path getPathFromClassName(String className) {
```

Comment:
Should prob have some tests for this method. For example, what would happen if `className` was empty?